source/solvers/navier_stokes_base.cc

```diff
@@ -1721,137 +1737,289 @@ NavierStokesBase<dim, VectorType, DofsType>::establish_solid_domain(
+          if (non_zero_constraints)
+            {
+              this->nonzero_constraints.add_line(local_dof_indices[i]);
+              this->nonzero_constraints.set_inhomogeneity(local_dof_indices[i],
+                                                          0);
+            }
+          else
+            zero_constraints.add_line(local_dof_indices[i]);
+        }
+    }
+}
+
+template <int dim, typename VectorType, typename DofsType>
+void
+NavierStokesBase<dim, VectorType, DofsType>::flag_dofs_in_solid(
+  std::unordered_set<types::global_dof_index> &dofs_are_in_solid,
+  const std::vector<types::global_dof_index>  &local_dof_indices)
+{
+  for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+    {
+      const unsigned int component =
+        this->fe->system_to_component_index(i).first;
+      if (component == dim)
+        {
+          dofs_are_in_solid.insert(local_dof_indices[i]);
+        }
+    }
+}
+
+template <int dim, typename VectorType, typename DofsType>
+bool
+NavierStokesBase<dim, VectorType, DofsType>::check_cell_is_in_solid(
+  const std::unordered_set<types::global_dof_index> &dofs_are_in_solid,
+  const std::vector<types::global_dof_index>        &local_dof_indices)
+{
+  for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+    {
+      auto search = dofs_are_in_solid.find(local_dof_indices[i]);
+      if (search != dofs_are_in_solid.end())
+        return true;
+    }
+  return false;
+}
+
+template <int dim, typename VectorType, typename DofsType>
+void
+NavierStokesBase<dim, VectorType, DofsType>::flag_dofs_connected_to_fluid(
+  std::unordered_set<types::global_dof_index> &dofs_are_connected_to_fluid,
+  const std::vector<types::global_dof_index>  &local_dof_indices)
+{
+  for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+    {
+      const unsigned int component =
+        this->fe->system_to_component_index(i).first;
+      if (component == dim)
+        {
+          dofs_are_connected_to_fluid.insert(local_dof_indices[i]);
+        }
+    }
+}
+
+template <int dim, typename VectorType, typename DofsType>
+bool
+NavierStokesBase<dim, VectorType, DofsType>::check_cell_is_connected_to_fluid(
+  const std::unordered_set<types::global_dof_index>
+                                             &dofs_are_connected_to_fluid,
+  const std::vector<types::global_dof_index> &local_dof_indices)
+{
+  for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+    {
+      auto search = dofs_are_connected_to_fluid.find(local_dof_indices[i]);
+      if (search != dofs_are_connected_to_fluid.end())
+        return true;
+    }
+  return false;
+}
+
+template <int dim, typename VectorType, typename DofsType>
+void
+NavierStokesBase<dim, VectorType, DofsType>::constrain_pressure(
+  const bool                                 &non_zero_constraints,
+  const std::vector<types::global_dof_index> &local_dof_indices,
+  AffineConstraints<double>                  &zero_constraints)
```

Comment:
Since this function is called within a cell, I would consider inlining it and defining it in Utilities or somewhere else